pilot/pkg/model/context.go

```diff
@@ -849,14 +854,23 @@ func (node *Proxy) SetServiceInstances(serviceDiscovery ServiceDiscovery) {
 	node.ServiceInstances = instances
 }
 
-// SetWorkloadLabels will set the node.Metadata.Labels only when it is nil.
+// SetWorkloadLabels will set the node.Metadata.Labels.
+// It merges both node meta labels and workload labels and give preference to workload labels.
+// Note: must be called after `SetServiceInstances`.
 func (node *Proxy) SetWorkloadLabels(env *Environment) {
-	// First get the workload labels from node meta
-	if len(node.Metadata.Labels) > 0 {
-		return
+	labels := env.GetProxyWorkloadLabels(node)
+	if len(labels) > 0 {
```

Comment:
Not possible a pod has 0 labels in istio, at least some labels are injected by istio

Reply:
Not all pods are injected

Reply:
The pod is originating the xds req, isn't it injected with a sidecar?

Reply:
Not necessarily, could be a proxy less grpc. Or gateway, or istioctl kube-inject, or maybe custom injection template.

Tbh it's probably very rare edge case

Reply:
For safety, we donot need to consider this labels = 0, in reality, no CICD will have no labels.

Reply:
should we just have a bool returned with the map to indicate if it was set or not? That makes things a bit more explicit, and should be a simple addition. Even though the edge case is rare it seems like it should be handled